src/stopwatch.hpp

```diff
@@ -0,0 +1,143 @@
+  double Elapsed(TimeScale scale = TimeScale::no_scale) {
+    scale = ((scale == TimeScale::no_scale) ? scale_ : scale);
+    stop_time_ = GetCurrentTime();
+    double elapsed_seconds = DurationToSeconds(start_time_, stop_time_, scale);
+    return elapsed_seconds;
+  };
+
+  // Returns the time elapsed since last lap (or start) and appends that to laps.
+  double Lap() {
+    time_point now = GetCurrentTime();
+    size_t num_laps = lap_times_.size();
+    time_point last_lap = ((num_laps == 0) ? start_time_ : lap_times_[num_laps - 1]);
+    lap_times_.push_back(now);
+    double lap_seconds = DurationToSeconds(last_lap, now);
+    lap_seconds_.push_back(lap_seconds);
+    return lap_seconds;
+  };
+  std::vector<double> &GetLaps() { return lap_seconds_; }
+
+  // Clears all total and lap times memory.
+  void Clear() {
+    last_subtotal_ = 0.0;
+    total_ = 0.0;
+    lap_times_.clear();
+    lap_seconds_.clear();
+  }
+
+  // Scale determines what times will be reported in: seconds, milliseconds, or
+  // nanoseconds.
+  void SetScale(TimeScale scale) { scale_ = scale; }
+  TimeScale GetScale() { return scale_; }
+
+  // Puts thread to sleep for given milliseconds.
+  static void Sleep(uint64_t wait_time) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(wait_time));
+  };
+
+  // Stopwatch functionality test.
+  static void UnitTest() {
+    std::cout << "Stopwatch Unit Test:" << std::endl;
+    Stopwatch watch(TimeScale::milliseconds);
+    Stopwatch::Sleep(10);
+    watch.Start();
+    Stopwatch::Sleep(10);
+    watch.Stop();
+    auto subtotal_1 = watch.GetLatestSubtotal();
+    Stopwatch::Sleep(10);
+    watch.Start();
+    Stopwatch::Sleep(10);
+    watch.Lap();
+    Stopwatch::Sleep(20);
+    watch.Lap();
+    auto subtotal_2 = watch.Stop();
+    auto laps = watch.GetLaps();
+    auto total = watch.GetTotal();
+    std::vector<double> totals = {subtotal_1, subtotal_2, total};
+    std::cout << "Laps (10ms,20ms): " << laps << ", Total (10ms,30ms,40ms): " << totals
+              << std::endl;
+  };
+
+ private:
+  time_point GetCurrentTime() { return clock::now(); }
+
+  double DurationToSeconds(time_point &start_time, time_point &end_time,
+                           TimeScale scale = TimeScale::no_scale) {
+    scale = ((scale == TimeScale::no_scale) ? scale_ : scale);
+    auto time_passed = end_time - start_time;
+    auto nanosec = double(
+        std::chrono::duration_cast<std::chrono::nanoseconds>(time_passed).count());
+    double seconds;
+
+    switch (scale) {
+      case second_scale:
+        seconds = nanosec / nanosec_per_sec;
+        break;
+      case millisecond_scale:
+        seconds = nanosec / nanosec_per_millisec;
+        break;
+      case nanosecond_scale:
+        seconds = nanosec;
+        break;
+    }
+    return seconds;
+  };
+
+  TimeScale scale_;
+  time_point start_time_;
+  time_point stop_time_;
+  bool is_running_;
+  double latest_subtotal_;
```

Comment:
Is this data member necessary?

Reply:
I also added a description for usage to the top of the file.

Reply:
These are all good improvements, but I guess my original question stands: is this the same as the sum of the lap vector? If so, can't we eliminate this data member?